### PR TITLE
Gh1169 Prevent flatrate terms from completing if overlapping

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/1169

### CODE REVIEW

File: de.metas.contracts/src/main/java/de/metas/flatrate/process/C_Flatrate_Term_Create.java [hunk 1]
```diff
@@ -122,24 +126,45 @@ private I_C_Flatrate_Term createTerm(final I_C_BPartner partner)
...
-			term.setEndDate(endDate);
+			newTerm.setEndDate(endDate);
+		}
+
+		InterfaceWrapperHelper.save(newTerm);
+
+		// task #1169
+		// in case the term to be created will overlap with other term regarding time period and product
+		// the user must be announced about this and the new term shall not be completed
+		final boolean hasOverlappingTerms = flatrateBL.hasOverlappingTerms(newTerm);
```
Comment (metas-ts) [inline de.metas.contracts/src/main/java/de/metas/flatrate/process/C_Flatrate_Term_Create.java]: Please make this check not in this process, but in a model interceptor.
Goal: the check shall also be made if a term is completed without this process taking place

File: de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java [hunk 2]
```diff
@@ -1737,4 +1706,193 @@ public void voidIt(final I_C_Flatrate_Term term)
...
+		for (final I_C_Flatrate_Term term : terms)
+		{
+
+			// Only consider completed terms
+			if (!X_C_Flatrate_Term.DOCSTATUS_Completed.equals(term.getDocStatus()))
+			{
+				continue;
+			}
+			
+			if (periodsOverlap(newTerm, term))
```
Comment (metas-ts) [inline de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java]: here i aggree with [codacy](https://www.codacy.com/app/metasfresh/metasfresh/pullRequest?prid=591574): nesting those two ifs doesn't improve readability. pls combine them

File: de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java [hunk 3]
```diff
@@ -1737,4 +1706,193 @@ public void voidIt(final I_C_Flatrate_Term term)
...
+
+			for (final I_C_Flatrate_Matching matching : flatrateMatchings)
+			{
+				if (newProduct.getM_Product_ID() == matching.getM_Product_ID())
+				{
+					// there is one matching for the same product as the one from the new term
+					return true;
+				}
+
+				if (matching.getM_Product() == null)
```
Comment (metas-ts) [inline de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java]: same comment as [further up](https://github.com/metasfresh/metasfresh/pull/1270/files#r109620192)

File: de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java [hunk 4]
```diff
@@ -1737,4 +1706,193 @@ public void voidIt(final I_C_Flatrate_Term term)
...
+
+			for (final I_C_Flatrate_Matching matching : flatrateMatchings)
+			{
+				if (product.getM_Product_ID() == matching.getM_Product_ID())
+				{
+					// there is one matching for the same product as the one from the second term
+					return true;
+				}
+
+				if (matching.getM_Product() == null)
```
Comment (metas-ts) [inline de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java]: same comment as [further up](https://github.com/metasfresh/metasfresh/pull/1270/files#r109620192)

File: de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java [hunk 5]
```diff
@@ -1737,4 +1706,193 @@ public void voidIt(final I_C_Flatrate_Term term)
...
+				{
+					for (final I_C_Flatrate_Matching flatrateMatching : flatrateMatchings)
+					{
+						if (newFMProduct.getM_Product_ID() == flatrateMatching.getM_Product_ID())
+						{
+							// matchings with the same product
+							return true;
+						}
+
+						if (flatrateMatching.getM_Product() == null)
```
Comment (metas-ts) [inline de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java]: same :-)

File: de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java [hunk 6]
```diff
@@ -1737,4 +1706,193 @@ public void voidIt(final I_C_Flatrate_Term term)
...
+				// product is null. Check the product category
+				else
+				{
+					final org.compiere.model.I_M_Product_Category newFMProductCategory = newFlatrateMatching.getM_Product_Category_Matching();
+
+					for (final I_C_Flatrate_Matching flatrateMatching : flatrateMatchings)
+					{
+						final org.compiere.model.I_M_Product matchingProduct = flatrateMatching.getM_Product();
+
+						if (flatrateMatching.getM_Product() != null)
```
Comment (metas-ts) [inline de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java]: same :-)

File: de.metas.contracts/src/main/java/de/metas/flatrate/process/C_Flatrate_Term_Create.java (hunk 1 above)
Comment (metas-rc) [inline de.metas.contracts/src/main/java/de/metas/flatrate/process/C_Flatrate_Term_Create.java]: Done

File: de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java (hunk 2 above)
Comment (metas-rc) [inline de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java]: Done

File: de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java (hunk 3 above)
Comment (metas-rc) [inline de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java]: Done

File: de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java (hunk 4 above)
Comment (metas-rc) [inline de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java]: Done

File: de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java (hunk 5 above)
Comment (metas-rc) [inline de.metas.contracts/src/main/java/de/metas/flatrate/api/impl/FlatrateBL.java]: Done

**Pull request merged**